skimage/segmentation/_join.py

```diff
@@ -33,15 +39,34 @@ def join_segmentations(s1, s2):
     array([[0, 1, 3, 2],
            [0, 5, 3, 2],
            [4, 5, 5, 3]])
+    >>> join_segmentations(s1, s2, return_mapping=True)
+    (
+        array([[0, 1, 3, 2],
+               [0, 5, 3, 2],
+               [4, 5, 5, 3]]),
+        ArrayMap(array([0, 1, 2, 3, 4, 5]), array([0, 0, 1, 1, 2, 2])),
+        ArrayMap(array([0, 1, 2, 3, 4, 5]), array([0, 1, 0, 1, 0, 1]))
+    )
     """
     if s1.shape != s2.shape:
         raise ValueError("Cannot join segmentations of different shape. "
                          f"s1.shape: {s1.shape}, s2.shape: {s2.shape}")
-    s1 = relabel_sequential(s1)[0]
-    s2 = relabel_sequential(s2)[0]
-    j = (s2.max() + 1) * s1 + s2
-    j = relabel_sequential(j)[0]
-    return j
+    # Reindex input label images
+    s1, _, backward_map1 = relabel_sequential(s1)
+    s2, _, backward_map2 = relabel_sequential(s2)
+    # Create joined label image
+    factor = s2.max() + 1
+    j = factor * s1 + s2
+    labels_joined = np.unique(j)
```

Comment:
The original code was reassigning modified data to the same references.
Here, `labels_joined = np.unique(j)` must be computed before relabeöing `j`.

- Usually I would use unique reference names to avoid depending on the correct order of lines (and potential breakages). I don't know whether `j`was overwritten for performance/memory reasons.
- Also, the unique operation is only needed when return_mapping is True. I don't know what performance impact it has when running the function with its defaults. One could move `labels_joined` into a conditional, but then the reference is potentially undefined, unless we also define it in an else branch (`else: labels_joined = None`)

Reply:
> Usually I would use unique reference names to avoid depending on the correct order of lines (and potential breakages). I don't know whether `j` was overwritten for performance/memory reasons.

Nope! It was rewritten willy-nilly. 😅 I wrote this code ages ago when I was much less experienced. I 100% agree about not reusing reference names and support changing the first reference to e.g. `j_initial`? I'm also ok with using more descriptive names throughout.

> Also, the unique operation is only needed when return_mapping is True. I don't know what performance impact it has when running the function with its defaults.

But if you don't overwrite `j`, you can move it to after the return clause and all is well, right?

Reply:
Yes, I renamed the variables (hopefully not too verbose) and moved the line after the return.